### PR TITLE
Issue 2316 - Image pull does not use the auth for docker.io repo

### PR DESCRIPTION
**Pull request by linggao**
Signed-off-by: linggao <linggao@us.ibm.com>

### CODE REVIEW

File: imagefetch/imagepull.go [hunk 1]
```diff
@@ -116,21 +115,40 @@ func pullImageFromRepos(config config.Config, authConfigs map[string][]docker.Au
...
-				} else if i < len(auth_array)-1 {
-					glog.V(5).Infof("Docker image pull(s) failed for service %v docker image %v with auth name %v. Error: %v. Try next auth.", name, service.Image, auth.Username, err)
-				}
+			domain = "docker.io"
+		}
+
+		// get all the auths for this domain or repo.
+		auth_array := []docker.AuthConfiguration{}
+		for k, _ := range authConfigs {
+			a_auth := authConfigs[k]
```
Comment (dabooz) [inline imagefetch/imagepull.go:126]: Why not:
 for k, a_auth := range authConfigs{ ?

File: imagefetch/imagepull.go [hunk 2]
```diff
@@ -116,21 +115,40 @@ func pullImageFromRepos(config config.Config, authConfigs map[string][]docker.Au
...
+		}
+
+		// try auths one at a time
+		var err error
+		for i, auth := range auth_array {
+			err = pullSingleImageFromRepo(client, opts, auth)
+			if err == nil {
+				break
+			} else if i < len(auth_array)-1 {
+				glog.V(5).Infof("Docker image pull(s) failed for service %v docker image %v with auth name %v. Error: %v. Try next auth.", name, service.Image, auth.Username, err)
```
Comment (dabooz) [inline imagefetch/imagepull.go:142]: Why not put out a log entry for the last element in the auth_array?

**Review by dabooz — CHANGES_REQUESTED** [review]

File: imagefetch/imagepull.go (hunk 1 above)
Comment (linggao) [reply imagefetch/imagepull.go:126]: I worry about the a_auth changes value as we iterates. I forgot what was the trick for this.

Comment (dabooz) [reply imagefetch/imagepull.go:126]: The problem with iteration occurs when you take the address of an element and try to use it later. In this case you're appending the array element to another array, so I dont think you have that problem. See this example: https://play.golang.org/p/lc8SzA3LVGx that *I think* verifies that there is no problem.

I think you can use authConfig[k] in the append where you use a_auth right now.

**Review by dabooz — APPROVED** [review]

**Pull request merged**